### ISSUE TITLE
Popup outside axes/figure

### ISSUE THREAD

**Issue by AgilentGCMS** [joferkington/mpldatacursor#28]
Hello,

Is it possible to control the position of the popup instead of having it always above and to the left of the pointer/data point? With the current setup, the popup corresponding to points near the left/top edge of a figure can be cut off because it's outside the figure (see attached example).

Thanks

![datacursor_problem](https://cloud.githubusercontent.com/assets/3866945/6117008/9ce5843a-b06e-11e4-8950-dc07b1dc9d39.png)

**Comment by joferkington** [comment]
You can control the position of the cursor through the `xytext` kwarg (it's passed on to `annotate`), but it's static. It won't automatically change when the "popup" goes off the edge of the figure.

I should add an option to make the "popup" stay more-or-less within the bounds of the axes, though.  Thanks for the suggestion!  I'll try to get to it this weekend or the next.

**Comment by AgilentGCMS** [comment]
Hi,
Yes, I discovered the `xytext` option, but as you say, it's static. I made some modifications to datacursor.py based on my understanding of the script, which does the job (patch below).
```
--- /home/i/Downloads/mpldatacursor-master/mpldatacursor/datacursor.py  2014-07-15 17:56:22.000000000 -0600
+++ /usr/local/lib/python2.7/dist-packages/mpldatacursor-0.5_dev-py2.7.egg/mpldatacursor/datacursor.py  2015-02-10 13:14:13.382363988 -0700
@@ -44,5 +44,5 @@
                 textcoords='offset points', 
                 bbox=dict(boxstyle='round,pad=0.5', fc='yellow', alpha=0.5),
-                arrowprops=dict(arrowstyle='->', connectionstyle='arc3,rad=0'))
+                arrowprops=dict(arrowstyle='-|>', connectionstyle='arc3,rad=0'))
 
     default_keybindings = dict(hide='d', toggle='t')
@@ -259,5 +259,4 @@
             for ann in self.annotations.values():
                 ann.set_visible(False)
-
         self.update(event, annotation)
 
@@ -372,4 +371,9 @@
         # changing of relative position of text box...)
         dx, dy = kwargs['xytext']
+        # SBi
+        # Store the default dx and dy magnitudes
+        self.dx_abs = abs(dx)
+        self.dy_abs = abs(dy)
+        # SBf
         horizontal = {1:'left', 0:'center', -1:'right'}[np.sign(dx)]
         vertical = {1:'bottom', 0:'center', -1:'top'}[np.sign(dy)]
@@ -471,4 +475,18 @@
         annotation.set_text(self.formatter(**info))
         annotation.xy = info['x'], info['y']
+        # SBi
+        x, y = info['x'], info['y']
+        xmid = np.average(event.artist.axes.get_xlim())
+        ymid = np.average(event.artist.axes.get_ylim())
+        # Which side of the midpoint is the data point?
+        dx = self.dx_abs if x < xmid else -self.dx_abs
+        dy = self.dy_abs if y < ymid else -self.dy_abs
+        annotation.xytext = (dx,dy)
+
+        horizontal = {1:'left', 0:'center', -1:'right'}[np.sign(dx)]
+        vertical = {1:'bottom', 0:'center', -1:'top'}[np.sign(dy)]
+        annotation.set_ha(horizontal)
+        annotation.set_va(vertical)
+        # SBf
 
         # In case it's been hidden earlier...
```
The `+++` file is the one with my modifications. This works for me, but I haven't checked all the corner cases. Hope this helps.

**Comment by joferkington** [comment]
Thanks for this, and sorry for my delay in replying.  I'll add it (or something very similar) soon and credit you in the changelog.

**Comment by joferkington** [comment]
Implemented in 1161a17852e44a900e8b904f871653968a0ec650

Instead of trying to keep things inside the axes, it tries to stay visible (i.e. inside the figure) instead.  

Thanks!
